andela/knights-bn-backend
# #170947678 User can edit an open request

#### What does this PR do?
- This PR fixes the way the user requests a two-way trip and enables him/her to edit a trip request that is still open (or pending).
#### Description of Task to be completed?
- Fix a two-way requesting process by checking if a user has an admin. 
ENDPOINT: ```POST http://localhost:4000/api/v1/trips/returnTrip```
- Create an endpoint to the edit request.
ENDPOINT: ```PATCH http://localhost:4000/api/v1/trips/edit/?requestId```

- Test the two endpoints.
- Document the endpoint.
#### How should this be manually tested?
- open your terminal and run this command `git clone https://github.com/andela/knightsbn    backend.git` to clone repository
- run `cd knights-bn-backend` to get inside the project
- run `git fetch && git checkout ft-edit-open-request-170947678` 
- run `npm install` to install all dependencies
- check if you have PostgresQL database if you don't have it install it from here [PostgreSQL](https://www.2ndquadrant.com/en/blog/pginstaller-install-postgresql/)
- create a Postgres database locally.
- create a .env file, paste in below variables and replace xxxx with the correct values: (Check code-knights `slack channel` to find the correct values)

```
DB_USER='name of the user of the database'
DB_PASS='password of your Postgres'
DB_NAME='name of the database || Barefoot'
DB_TEST_NAME='name of the database for test || Barefoot_test'
NODE_ENV='development'
SECRETKEY='PrivateKey'
GOOGLE_CLIENT_ID = 'XXXXX'
GOOGLE_CLIENT_SECRET = 'XXXXX'
GOOGLE_CALLBACK_URL = 'XXXXX'
FCBK_CLIENT_ID = 'XXXXX'
FCBK_CLIENT_SECRET = 'XXXXX'
FCBK_CALLBACK_URL = 'XXXXX'
BN_API_KEY= 'XXXX'
MANAGER_PSWD= 'XXXX'
```
- run `npm run createTables` to create the database tables and seed in primary data.
- run `npm dev-start` to start the server.
- request a two-way trip: ```POST http://localhost:4000/api/v1/trips/returnTrip```

1. Login first

Login info: 
```
{
    "email": "ishimwe@gmail.com",
    "password": "Password@1"
}
```
![image](https://user-images.githubusercontent.com/30593186/76017765-cfa0b300-5f27-11ea-9cef-18771a8b357e.png)

2. Create a trip

- edit a trip: ```POST http://localhost:4000/api/v1/trips/returnTrip```

Trip info
```
{
	"origin": "Kigali",
	"destination": "Kampala",
	"departureDate": "2020-04-01",
	"returnDate": "2020-04-10",
	"reason": "Having fun",
	"accommodation": "Z campus"
}
```

![image](https://user-images.githubusercontent.com/30593186/76017887-07a7f600-5f28-11ea-90fc-701f26fcbbc1.png)

3. Update the trip

Edits: 
```
{
	"destination": "Bujumbura",
        "reason" : "meetup"
}
```
![image](https://user-images.githubusercontent.com/30593186/76023101-2c549b80-5f31-11ea-9072-aa446478d82e.png)

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#170947678](https://www.pivotaltracker.com/story/show/170947678)

#### Screenshots
1. validations : 

- Origin and destination have to differ

![image](https://user-images.githubusercontent.com/30593186/76023538-07145d00-5f32-11ea-9c14-66664a7df84c.png)

- Departure date has to be set before the return data.

![image](https://user-images.githubusercontent.com/30593186/76023848-af2a2600-5f32-11ea-80a9-75e941aed9be.png)

- Departure date has to be set from today's date

![image](https://user-images.githubusercontent.com/30593186/76023899-cb2dc780-5f32-11ea-8f2b-3761b0020182.png)

- Can't create a new request trip that collides with already requested trip

![image](https://user-images.githubusercontent.com/30593186/76024020-0c25dc00-5f33-11ea-8dff-d557c33e3f06.png)

- While editing a request, an edit can't be empty
![image](https://user-images.githubusercontent.com/30593186/76024367-a25a0200-5f33-11ea-9157-2ff4b9005736.png)
